agent/lib/kontena/service_pods/lifecycle_hook_manager.rb

```diff
@@ -0,0 +1,159 @@
+          end
+        end
+        true
+      rescue => exc
+        log_service_pod_event("service:create_instance", exc.message, Logger::ERROR)
+        error exc.message
+        false
+      end
+
+      # @param [Docker::Container] service_container
+      # @return [Boolean]
+      def on_pre_stop(service_container)
+        hooks_for('pre_stop').each do |hook|
+          info "running pre_stop hook: #{hook['cmd']}"
+          command = build_cmd(hook['cmd'])
+          log_hook_output(service_container.id, "running pre_stop hook: #{hook['cmd']}", 'stdout')
+          _, _, exit_code = service_container.exec(command) { |stream, chunk|
+            log_hook_output(service_container.id, chunk, stream)
+          }
+          if exit_code != 0
+            raise "Failed to execute pre_stop hook: #{hook['cmd']} (exit code: #{exit_code}"
+          end
+        end
+        true
+      rescue => exc
+        log_service_pod_event("service:create_instance", exc.message, Logger::ERROR)
+        error exc.message
+        false
+      end
+
+      # @param [String] type
+      # @return [Array<Hash>]
+      def hooks_for(type)
+        hooks = service_pod.hooks.select { |h|
+          h['type'] == type && !h['oneshot'] || (h['oneshot'] && !oneshot_cache.include?(h))
+        }
+        hooks.each do |h|
+          if h['oneshot'] && !oneshot_cache.include?(h)
+            mark_oneshot_hook(h)
+          end
+        end
+
+        hooks
+      end
+
+      # @param [String] cmd
+      # @return [Array<String>]
+      def build_cmd(cmd)
+        command = ['/bin/sh', '-c', cmd]
+
+        command
+      end
+
+      # @param service_pod [ServicePod]
+      # @param cmd [String]
+      # @return [Hash]
+      def config_container(service_pod, cmd)
+        service_pod.entrypoint = ['/bin/sh', '-c']
+        service_pod.cmd = cmd
+        service_config = super(service_pod)
+        service_config['HostConfig'].delete('RestartPolicy')
+        service_config['Labels']['io.kontena.container.type'] = 'service_hook'
+
+        service_config
+      end
+
+      # @param [String] id
+      # @param [String] line
+      # @param [String] type
+      def log_hook_output(id, line, type)
+        data = [{
+          id: id,
+          time: Time.now.utc.xmlschema,
+          type: type,
+          data: line
+        }]
+        rpc_client.async.notification('/containers/log_batch', [data])
+      end
+
+      # @param [String] type
+      # @param [String] data
+      # @param [Integer] severity
+      def log_service_pod_event(type, data, severity = Logger::INFO)
+        super(service_pod.service_id, service_pod.instance_number, type, data, severity)
+      end
+
+      # @param hook [Hash]
+      def mark_oneshot_hook(hook)
+        pod = {
+          service_id: service_pod.service_id,
+          instance_number: service_pod.instance_number
+        }
+        rpc_client.async.request('/node_service_pods/mark_oneshot_hook', [node.id, pod, hook])
```

Comment:
> Just thinking should we remove async from there.. then if RPC fails it would not mark hook as done in cache.

I didn't notice that this was `async`. IMO yes, it's safer to make this sync, because that means that errors cause oneshot hooks to never get executed, rather than having them get executed multiple times.